File: AddTwoNumbers/AddTwoNumbers.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */

function ListNode(val) {
    this.val = val;
    this.next = null;
}
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */

var addTwoNumbers = function(l1, l2) {
    let n1 = l1;
    let n2 = l2;
    let a1 = n1.val;
    let a2 = n2.val;
    let carry = 0;

    let head = new ListNode(-1);
    let next = head;

    while(n1 != null || n2 != null || carry != 0)
    {
        let i = a1 + a2 + carry;
        carry = (i/10) | 0;
        i = i % 10;
        next.next = new ListNode(i);
        next = next.next;
        if(n1 != null)
        {
            n1 = n1.next;
            a1 = (n1 != null) ? n1.val : 0;
        }
        if(n2 != null)
        {
            n2 = n2.next;
            a2 = (n2 != null) ? n2.val : 0;
        }
    }
    return head.next;
};

/*
Equal to or fasther than 97.33% of Javascript submissions on LeetCode
https://leetcode.com/submissions/detail/144990026/
*/

File: AddTwoNumbers/AddTwoNumbers.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def addTwoNumbers(self, l1, l2):
        n1 = l1
        n2 = l2
        a1 = n1.val
        a2 = n2.val
        carry = 0

        head = ListNode(-1)
        nextNode = head

        while(n1 is not None or n2 is not None or carry != 0):
            i = a1 + a2 + carry
            carry = int(i/10)
            i = i % 10
            nextNode.next = ListNode(i)
            nextNode = nextNode.next
            if(n1 is not None):
                n1 = n1.next
                if(n1 is not None):
                    a1 = n1.val
                else:
                    a1 = 0
            if(n2 is not None):
                n2 = n2.next
                if(n2 is not None):
                    a2 = n2.val
                else:
                    a2 = 0
        return head.next

#
# Equal to or fasther than 94.01% of Python3 submissions on LeetCode
# https://leetcode.com/submissions/detail/144990026/
#